### ISSUE TITLE
Profiler silently drops GPU kernel events at the start of each profiling session; loss grows with accumulated GPU idle time

### ISSUE THREAD

**Issue by bigSheep123** [pytorch/pytorch#192021]
### 🐛 Describe the bug

We use `torch.profiler` in a kernel benchmarking harness to get kernel-only timings: each timed call is wrapped in `record_function`, and we read the projected GPU-side user annotations back from the kineto trace. In long-running benchmark processes we kept seeing sessions where some annotations never got projected onto the GPU timeline. It was always the first iterations of a session (49/50 windows, then 48/50, ... down to 0/50), and the longer the process had been alive, the more iterations were affected.

After digging through libkineto we believe this is what happens:

- The GPU kernel activity records for those iterations do reach libkineto, but they carry timestamps that lie *before* `captureWindowStartTime_`, so `outOfRange()` drops them (`GenericActivityProfiler.cpp:329`/`413` on kineto main; `CuptiActivityProfiler.cpp:547` in the pin shipped with torch 2.10).
- The GPU user annotation for an iteration is only created while handling its kernel record, and `handleGpuActivity` checks `outOfRange` before anything else, so the annotation for that iteration silently disappears.
- The only visible trace of the drop is the `Record counts: Out-of-range = N` INFO log line. Nothing in the trace file or the python API tells you the trace is incomplete.

Evidence that the records are delivered and then dropped, rather than lost in CUPTI: for failing sessions the "Processed N GPU records" count is exactly equal to that of identical passing sessions, and the Out-of-range counter matches the number of missing iterations one for one.

The timestamp error itself grows with accumulated GPU *idle* time — about 6.4 us/s on our machine (H200, SM clocks locked, persistence mode on) — and is reset by GPU activity. Only the first records of each session are affected; later records in the same session line up with their launch records to within a few microseconds again. Our best guess is that CUPTI's mapping of GPU timestamps into the registered timestamp callback domain (kineto registers rdtsc via `cuptiActivityRegisterTimestampCallback`) goes stale across idle periods and is refreshed shortly after collection starts. That part is opaque to us, but the effect on the kineto side is deterministic and easy to reproduce.

This bites hard in benchmark/CI processes: JIT compilation phases keep the GPU idle for minutes, and every profiler session that follows loses its head iterations. With enough accumulated idle time an entire 50-iteration session projects 0 windows.

Repro (any CUDA machine, ~2.5 min):

```python
import time
import torch
from torch.autograd.profiler import DeviceType

REGION = "timed_region"
N_REPEAT = 50

a = torch.randn(256, 256, device="cuda", dtype=torch.float16)
for _ in range(10):
    a @ a
torch.cuda.synchronize()

t0 = time.monotonic()
for round_idx in range(10):
    with torch.profiler.profile(
        activities=[
            torch.profiler.ProfilerActivity.CPU,
            torch.profiler.ProfilerActivity.CUDA,
        ],
    ) as prof:
        for _ in range(N_REPEAT):
            with torch.profiler.record_function(REGION):
                a @ a
            torch.cuda.synchronize()

    cpu_ids, gpu_ids = [], set()
    for evt in prof.profiler.kineto_results.events():
        if evt.is_user_annotation() and evt.name() == REGION:
            if evt.device_type() == DeviceType.CUDA:
                gpu_ids.add(evt.correlation_id())
            else:
                cpu_ids.append((evt.start_ns(), evt.correlation_id()))
    cpu_ids.sort()
    missing = [i for i, (_, c) in enumerate(cpu_ids) if c not in gpu_ids]
    print(
        f"t={time.monotonic() - t0:5.0f}s  "
        f"gpu_windows={len(gpu_ids)}/{N_REPEAT}  missing_repeats={missing}",
        flush=True,
    )
    time.sleep(15)  # GPU idle; the drift that causes the drops accumulates here
```

Output on our machine:

```
t=    1s  gpu_windows=50/50  missing_repeats=[]
t=   16s  gpu_windows=49/50  missing_repeats=[0]
t=   31s  gpu_windows=48/50  missing_repeats=[0, 1]
t=   46s  gpu_windows=47/50  missing_repeats=[0, 1, 2]
t=   61s  gpu_windows=46/50  missing_repeats=[0, 1, 2, 3]
t=   76s  gpu_windows=45/50  missing_repeats=[0, 1, 2, 3, 4]
t=   91s  gpu_windows=44/50  missing_repeats=[0, 1, 2, 3, 4, 5]
t=  106s  gpu_windows=43/50  missing_repeats=[0, 1, 2, 3, 4, 5, 6]
t=  121s  gpu_windows=42/50  missing_repeats=[0, 1, 2, 3, 4, 5, 6, 7]
t=  136s  gpu_windows=41/50  missing_repeats=[0, 1, 2, 3, 4, 5, 6, 7, 8]
```

With `KINETO_LOG_LEVEL=1` the libkineto log shows `Record counts: Out-of-range = 0, 1, 2, ...` increasing by one per idle period, matching the missing iterations exactly. Replacing the sleep with GPU work keeps everything at 50/50 indefinitely (we ran 400 sessions over ~9 minutes without a single drop), which is how we isolated the idle-time dependence.

Two things seem worth fixing on the kineto side regardless of where the timestamp error ultimately comes from:

1. Dropped-record counts are invisible to users. `ecs_.out_of_range_events` and friends only go to an INFO log. Exporting them through trace metadata / the profiler result would at least let callers detect that a trace is incomplete. I'd be happy to put together a PR for that.
2. `outOfRange()` compares CUPTI-domain timestamps against window bounds taken directly from `system_clock` (`startTraceInternal`), so any disagreement between the two clock chains turns into silent data loss. Tagging or clamping instead of dropping, or taking the window bounds in the same domain as the activities, would make this failure mode a lot less destructive.

Possibly related: pytorch/kineto#1354 fixed a different silent-drop path (CUPTI buffer ordering vs. correlation maps; not in the kineto pin of torch 2.10), and pytorch/kineto#996 looks like an extreme case of the same window filtering.

Workaround for anyone hitting this: run a profiler warmup step (`schedule(wait=0, warmup=1, active=1)`) or a couple of untimed dummy iterations right after entering the profiler. The stale-timestamp records then fall into the warmup phase and the active window is clean — with that change we get 50/50 windows even in processes where the plain pattern loses 20/50.

### Versions

PyTorch 2.10.0+cu129, CUDA 12.9, CUPTI 28, driver 575.57.08
NVIDIA H200, Linux 6.2.0-39-generic, Python 3.10.12

The relevant code paths (`outOfRange`, window bounds from `system_clock`, log-only drop counters) are unchanged on kineto main as of 70e46d9.

cc @robieta @chaekit @guotuofeng @guyang3532 @dzhulgakov @davidberard98 @briancoutinho @sraikund16 @sanrise @mwootton @divyanshk @jiannanWang @scotts @ryanzhang22

**Comment by bigSheep123** [comment]
Update: we wrote a standalone CUPTI-only repro (no torch, no kineto) to make sure the corruption really originates below libkineto, and it pinned the trigger more precisely than the original report.

The trigger is **re-registering the timestamp callback**. libkineto calls `cuptiActivityRegisterTimestampCallback` in `CuptiActivityProfiler::configure()`, i.e. once per profiling session. In a plain C program that mimics this (register an rdtsc callback, enable RUNTIME + CONCURRENT_KERNEL, launch 50 kernels with a sync after each, disable, forced flush, repeat), CUPTI starts handing back kernel records whose timestamps are not in the callback domain at all — off by ~1e14 ns, which looks like a fall-back to CUPTI's default clock. The runtime records in the same buffer stay correct, so kernel records claim to start long before their own `cudaLaunchKernel` began. The number of corrupted head records grows by exactly one per re-registration and accumulates across the process:

| callback registration | between rounds | corrupted head kernel records |
|---|---|---|
| once per process | 15 s idle | 0 across all rounds |
| re-registered every round (what kineto does per session) | 15 s idle | round k has exactly k |
| re-registered every round | GPU busy | round k has exactly k |

So the idle-time correlation in my first post is a property of how this shows up under torch (still not fully explained), not the root trigger. Registering the callback once keeps everything clean even with idle gaps and enable/disable cycling.

Two consequences:

1. There is a one-line-ish mitigation available in libkineto: register the timestamp callback once per process (e.g. behind a `std::once_flag`) instead of on every `configure()`. I can send a PR if that sounds reasonable.
2. The corrupted-timestamp behavior itself is CUPTI-internal (reproduces with nothing but the CUPTI API); we'll report it to NVIDIA separately. CUPTI 28 / CUDA 12.9 / driver 575.57.08 / H200.

<details>
<summary>Standalone repro (nvcc -O2 raw_cupti_probe.cu -lcupti; run with GPU idle or busy between rounds, same result)</summary>

```cpp
// Standalone CUPTI probe, no torch/kineto involved. Registers a timestamp
// callback, then per round: enable RUNTIME + CONCURRENT_KERNEL, launch 50
// kernels (sync after each), disable, forced flush. Kernel records are
// matched to their cudaLaunchKernel runtime records by correlationId.
// A kernel record whose start lies before its own launch started, or before
// a timestamp taken right after enable, is physically impossible and means
// the record content itself is corrupted.
// argv: rounds idle_seconds reregister(0/1) busy(0/1)
#include <cupti.h>
#include <cuda_runtime.h>
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <ctime>
#include <unistd.h>
#include <x86intrin.h>
#include <algorithm>
#include <vector>
#include <map>

#define CUPTI_CHECK(call)                                                  \
  do {                                                                     \
    CUptiResult _st = call;                                                \
    if (_st != CUPTI_SUCCESS) {                                            \
      const char* err;                                                     \
      cuptiGetResultString(_st, &err);                                     \
      fprintf(stderr, "CUPTI error %s at line %d\n", err, __LINE__);       \
      exit(1);                                                             \
    }                                                                      \
  } while (0)

static uint64_t nowNs() {
  struct timespec ts;
  clock_gettime(CLOCK_MONOTONIC_RAW, &ts);
  return (uint64_t)ts.tv_sec * 1000000000ull + ts.tv_nsec;
}

// same as kineto: raw rdtsc as the CUPTI timestamp callback
static uint64_t tscCb() { return __rdtsc(); }

// rough TSC frequency estimate (GHz), only used to print tick deltas as us
static double tscGhz() {
  uint64_t t0 = __rdtsc(); uint64_t n0 = nowNs();
  usleep(200000);
  uint64_t t1 = __rdtsc(); uint64_t n1 = nowNs();
  return (double)(t1 - t0) / (double)(n1 - n0);
}

struct LaunchRec { uint64_t start, end; };
struct KernelRec { uint64_t start, end; };
static std::map<uint32_t, LaunchRec> g_launches;   // correlationId -> runtime
static std::map<uint32_t, KernelRec> g_kernels;    // correlationId -> kernel

static void CUPTIAPI bufReq(uint8_t** buffer, size_t* size, size_t* maxRecs) {
  *size = 8 * 1024 * 1024;
  *buffer = (uint8_t*)aligned_alloc(8, *size);
  *maxRecs = 0;
}

static void CUPTIAPI bufDone(CUcontext, uint32_t, uint8_t* buffer,
                             size_t /*size*/, size_t validSize) {
  CUpti_Activity* rec = nullptr;
  while (cuptiActivityGetNextRecord(buffer, validSize, &rec) == CUPTI_SUCCESS) {
    if (rec->kind == CUPTI_ACTIVITY_KIND_RUNTIME) {
      auto* r = (CUpti_ActivityAPI*)rec;
      if (r->cbid == CUPTI_RUNTIME_TRACE_CBID_cudaLaunchKernel_v7000)
        g_launches[r->correlationId] = {r->start, r->end};
    } else if (rec->kind == CUPTI_ACTIVITY_KIND_CONCURRENT_KERNEL ||
               rec->kind == CUPTI_ACTIVITY_KIND_KERNEL) {
      auto* k = (CUpti_ActivityKernel9*)rec;
      g_kernels[k->correlationId] = {k->start, k->end};
    }
  }
  free(buffer);
}

__global__ void busyKernel(float* p) {
  float v = p[threadIdx.x];
  for (int i = 0; i < 2000; ++i) v = v * 1.0001f + 0.5f;
  p[threadIdx.x] = v;
}

int main(int argc, char** argv) {
  int rounds = argc > 1 ? atoi(argv[1]) : 10;
  int idle_s = argc > 2 ? atoi(argv[2]) : 15;
  int rereg = argc > 3 ? atoi(argv[3]) : 1;
  int busy = argc > 4 ? atoi(argv[4]) : 0;
  const int n_repeat = 50;

  float* d = nullptr;
  cudaMalloc(&d, 256 * sizeof(float));
  busyKernel<<<1, 256>>>(d);  // warmup
  cudaDeviceSynchronize();

  CUPTI_CHECK(cuptiActivityRegisterCallbacks(bufReq, bufDone));
  double ghz = tscGhz();
  fprintf(stderr, "tsc ~%.3f GHz\n", ghz);

  CUPTI_CHECK(cuptiActivityRegisterTimestampCallback(tscCb));
  for (int r = 0; r < rounds; ++r) {
    if (rereg && r > 0)
      CUPTI_CHECK(cuptiActivityRegisterTimestampCallback(tscCb));  // re-register, like kineto's configure()
    g_launches.clear();
    g_kernels.clear();

    CUPTI_CHECK(cuptiActivityEnable(CUPTI_ACTIVITY_KIND_RUNTIME));
    CUPTI_CHECK(cuptiActivityEnable(CUPTI_ACTIVITY_KIND_CONCURRENT_KERNEL));
    uint64_t window_start = __rdtsc();

    for (int i = 0; i < n_repeat; ++i) {
      busyKernel<<<1, 256>>>(d);
      cudaDeviceSynchronize();
    }

    CUPTI_CHECK(cuptiActivityDisable(CUPTI_ACTIVITY_KIND_CONCURRENT_KERNEL));
    CUPTI_CHECK(cuptiActivityDisable(CUPTI_ACTIVITY_KIND_RUNTIME));
    CUPTI_CHECK(cuptiActivityFlushAll(CUPTI_ACTIVITY_FLAG_FLUSH_FORCED));

    // kernel.start - launch.START (negative = impossible), plus a count of
    // kernel records that predate the post-enable window start.
    std::vector<std::pair<uint64_t, double>> deltas;  // (launch_start, delta_us)
    int before_window = 0;
    for (auto& kv : g_kernels) {
      auto it = g_launches.find(kv.first);
      if (it == g_launches.end()) continue;
      double d_us = ((double)kv.second.start - (double)it->second.start) / 1e3 / ghz;
      deltas.push_back({it->second.start, d_us});
      if (kv.second.start < window_start) before_window++;
    }
    std::sort(deltas.begin(), deltas.end());
    double dmin = 1e18;
    for (auto& p : deltas) if (p.second < dmin) dmin = p.second;
    printf("round=%2d matched=%zu/%d  dvs_launch_start_us first=%8.1f "
           "second=%8.1f mid=%6.1f last=%6.1f min=%8.1f  "
           "kernels_before_window_start=%d\n",
           r, deltas.size(), n_repeat,
           deltas.empty() ? 0.0 : deltas.front().second,
           deltas.size() > 1 ? deltas[1].second : 0.0,
           deltas.size() > 25 ? deltas[25].second : 0.0,
           deltas.empty() ? 0.0 : deltas.back().second,
           dmin, before_window);
    fflush(stdout);
    if (busy) {
      uint64_t bend = nowNs() + (uint64_t)idle_s * 1000000000ull;
      while (nowNs() < bend) { busyKernel<<<1, 256>>>(d); cudaDeviceSynchronize(); }
    } else {
      sleep(idle_s);
    }
  }
  cudaFree(d);
  return 0;
}

```

</details>

**Comment by JosueT1212** [comment]
Hi,may i work on this,if it still needs to be done?

**Comment by bigSheep123** [comment]
yeah，if you want to do, you can do it. I just find this bug.  But I am also external developers , I am not sure pytorch's developers will accept it.

**Comment by ryanzhang22** [comment]
@bigSheep123 Thanks for the report. Can you try running your repro on CUDA 13.1+ and let me know the result? There is a bug fixed in CUPTI tied to CUDA 13.1 which states: "Fixed an issue that could cause invalid (zero) kernel timestamps when using the function cuptiActivityRegisterTimestampCallback to register a timestamp callback. This issue was introduced in the CUDA 12.6 Update 2 release." In my local testing I can repro with your example on 12.9, but not on 13.2.

https://docs.nvidia.com/cupti/release-notes/release-notes.html#updates-in-cuda-13-1

R.e. reporting incomplete traces -- ack, we should look into this.
